### PR TITLE
bpo-36832: add zipfile.Path

### PR DESCRIPTION
**Pull request by jaraco**
https://bugs.python.org/issue36832

### CODE REVIEW

**Comment by jaraco** [comment]
This is a work-in-progress, as there's not yet any documentation.

File: Lib/test/test_zipfile.py [hunk 1]
```diff
@@ -2392,5 +2394,121 @@ def test_extract_command(self):
...
+    zf = zipfile.ZipFile(data, "w")
+    zf.writestr("a.txt", b"content of a")
+    zf.writestr("b/c.txt", b"content of c")
+    zf.writestr("b/d/e.txt", b"content of e")
+    zf.filename = "abcde.zip"
+    return zf
+
+
+@contextlib.contextmanager
+def tempdir():
```
Comment (matrixise) [inline Lib/test/test_zipfile.py]: There is `@test.support.temp_dir` I think you could use it.

Comment (jaraco) [reply Lib/test/test_zipfile.py]: Good suggestion. I hadn't used any of the test suite infrastructure because I wanted this contribution to be runnable outside of the cpython repo (for backport purposes). But I agree - it would be best to use standard patterns here.

File: Doc/library/zipfile.rst [hunk 2]
```diff
@@ -456,6 +465,62 @@ The following data attributes are also available:
...
 
 
+.. _path-objects:
+
+Path Objects
+------------
+
+.. class:: Path(root, at='')
+
+   Construct a Path object from a root zipfile (which may be a
```
Comment (warsaw) [inline Doc/library/zipfile.rst]: This needs to be more explicit about the semantics and values for the `root` and `at` parameters.

File: Doc/library/zipfile.rst [hunk 3]
```diff
@@ -456,6 +465,62 @@ The following data attributes are also available:
...
+   Invoke :method:`ZipFile.open` on the current path. Accepts
+   the same arguments as :method:`ZipFile.open`.
+
+.. method:: Path.listdir()
+
+   Enumerate the children of the current directory.
+
+.. method:: Path.is_dir()
+
+   Return True if the current context references a directory.
```
Comment (warsaw) [inline Doc/library/zipfile.rst]: We generally mark up `True` and `False` with wrapped double backticks.

File: Lib/test/test_zipfile.py [hunk 4]
```diff
@@ -8,6 +8,8 @@
 import struct
 import zipfile
 import unittest
+import tempfile
+import shutil
```
Comment (warsaw) [inline Lib/test/test_zipfile.py]: It looks to me like the imports in this file started as being sorted alphabetically.  `unittest` breaks this, but I would suggest resorting these imports.

File: Lib/test/test_zipfile.py [hunk 5]
```diff
@@ -2392,5 +2394,121 @@ def test_extract_command(self):
                             with open(path, 'rb') as f:
                                 self.assertEqual(f.read(), zf.read(zi))
 
+
+consume = tuple
```
Comment (warsaw) [inline Lib/test/test_zipfile.py:2399]: Why is this here? Why not just use `tuple` in place?

File: Lib/zipfile.py [hunk 6]
```diff
@@ -13,6 +13,8 @@
 import struct
 import binascii
 import threading
+import posixpath
+import functools
```
Comment (warsaw) [inline Lib/zipfile.py]: Sigh. ;)

File: Lib/zipfile.py [hunk 7]
```diff
@@ -2102,6 +2104,138 @@ def _compile(file, optimize=-1):
...
+    >>> (b / 'missing.txt').exists()
+    False
+
+    Coersion to string:
+
+    >>> str(c)
+    'abcde.zip/b/c.txt'
+    """
+
+    __repr = "{self.__class__.__name__}({self.root.filename!r}, {self.at!r})"
```
Comment (warsaw) [inline Lib/zipfile.py:2172]: Should this be wrapped in `<` and `>`?

File: Lib/test/test_zipfile.py (hunk 5 above)
Comment (jaraco) [reply Lib/test/test_zipfile.py:2399]: In the external library, this function came from [more_itertools.consume](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.consume) (because I can't import it from the stdlib docs). I chose not to inline `tuple` here because I wanted to be explicit about the purpose of this function (consume an iterable), but didn't feel like copy/pasting the recipe. My main motivation for not using it in-place is I don't want someone to wonder "why not `list` or `dict.from_keys()` or `set()` or `all()`?", any of which would work just as well, but none of which are optimal. I'll add a comment.

File: Lib/zipfile.py (hunk 7 above)
Comment (jaraco) [reply Lib/zipfile.py:2172]: Hmm. [The docs](https://docs.python.org/3/reference/datamodel.html#object.__repr__) say:

> If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment). If this is not possible, a string of the form <...some useful description...> should be returned.

Since the repr meets the first possibility, I think that precludes the use of `<>`.

File: Doc/library/zipfile.rst (hunk 2 above)
Comment (jaraco) [reply Doc/library/zipfile.rst]: In c8c151113c, I added a section about the 'at' parameter. What do you think is missing about the `root` parameter?

File: Lib/test/test_zipfile.py (hunk 5 above)
Comment (jaraco) [reply Lib/test/test_zipfile.py:2399]: In c8c151113c, I added a comment to that object explaining its purpose. Let me know what you think.

Comment (warsaw) [reply Lib/test/test_zipfile.py:2399]: Sounds good, and thanks for adding the comment, but maybe it should be a comment and not a string? :)

File: Doc/library/zipfile.rst (hunk 2 above)
Comment (warsaw) [reply Doc/library/zipfile.rst]: Maybe just mark up the first mention of `root` so that it's clear you're talking about the parameter?

**Review by warsaw — APPROVED** [review]
